include/boost/process/v2/impl/pid.ipp

```diff
@@ -22,6 +65,603 @@ pid_type current_pid() {return ::GetCurrentProcessId();}
+    std::vector<pid_type> vec;
+    vec.resize(proc_listpids(PROC_PPID_ONLY, (uint32_t)pid, nullptr, 0));
+    if (proc_listpids(PROC_PPID_ONLY, (uint32_t)pid, &proc_info[0], sizeof(pid_type) * cntp))
+    {
+        ec = detail::get_last_error();
+        return {};
+    }
+    auto itr = std::partition(vec.begin(), vec.end(), [](pid_type pt) {return pt != 0;});
+    vec.erase(itr, vec.end());
+    std::reverse(vec.begin(), vec.end());
+    for (unsigned i = 0; i < vec.size(); i++)
+    {
+        if (proc_info[i] == 1 && pid == 0)
+        {
+            vec.insert(vec.begin(), 0);
+            break;
+        }
+    }
+    return vec;
+}
+
+#elif (defined(__linux__) || defined(__ANDROID__))
+
+std::vector<pid_type> all_pids(std::error_code & ec)
+{
+    std::vector<pid_type> vec;
+    DIR *proc = opendir("/proc");
+    if (!proc)
+    {
+        ec = detail::get_last_error();
+        return vec;
+    } 
+    struct dirent *ent = nullptr;
+    while ((ent = readdir(proc))) 
+    {
+        if (!isdigit(*ent->d_name))
+            continue;
+        vec.push_back(atoi(ent->d_name));
+    }
+    closedir(proc);
+    return vec;
+}
+
+std::vector<pid_type> parent_pid(pid_type pid, std::error_code & ec) {
+    std::vector<pid_type> vec;
+    char buffer[BUFSIZ];
+    sprintf(buffer, "/proc/%d/stat", pid);
+    FILE *stat = fopen(buffer, "r");
+    if (!stat)
+    {
+        ec = detail::get_last_error();
+        return vec;
+    } 
+    else
+    {
+        std::size_t size = fread(buffer, sizeof(char), sizeof(buffer), stat);
+        if (size > 0) 
+        {
+            char *token = nullptr;
+            if ((token = strtok(buffer, " "))) {
+                if ((token = strtok(nullptr, " "))) {
+                    if ((token = strtok(nullptr, " "))) {
+                        if ((token = strtok(nullptr, " "))) {
+                            pid_type ppid = (pid_type)strtoul(token, nullptr, 10);
+                            vec.push_back(ppid);
+                        }
+                    }
+                }
+            }
+            if (!token)
+            {
+                ec = detail::get_last_error();
+                return vec;
+            }
+        }
+        fclose(stat);
+    }
+    return vec;
+}
+
+std::vector<pid_type> child_pids(pid_type pid, std::error_code & ec) {
+    std::vector<pid_type> vec;
+    std::vector<pid_type> pids = all_pids(ec);
+    for (std::size_t i = 0; i < pids.size(); i++)
+    {
+        std::vector<pid_type> ppid = parent_pid(pids[i], ec);
+        if (!ppid.empty() && ppid[0] == pid)
+        {
+            vec.push_back(pids[i]);
+        }
+    }
+    return vec;
+}
+
+#elif defined(__FreeBSD__)
+
+std::vector<pid_type> all_pids(std::error_code & ec)
+{
+    std::vector<pid_type> vec;
+    int cntp = 0;
+    kinfo_proc *proc_info = kinfo_getallproc(&cntp);
+    if (proc_info) 
+    {
+        for (int i = 0; i < cntp; i++)
+            vec.push_back(proc_info[i].ki_pid);
```

Comment:
technically a data leak, if `push_back` throws, `free` won't get called.

Also: call `vec.reserve()` in all the examples please, so we only alloc once.

Reply:
A smart_ptr might also do - `unique_ptr` with a custom deleter will do the trick.

Reply:
Are you saying we should use `std::unique_ptr<pid_type>` instead of `std::vector<pid_type>`, correct?